### ISSUE TITLE
azurerm_storage_account is failing when azure policy is enabled on storage account

### ISSUE THREAD

**Issue by krupakar1329** [hashicorp/terraform-provider-azurerm#25792]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment and review the [contribution guide](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/README.md) to help.

### Terraform Version

1.6.5

### AzureRM Provider Version

3.101.0

### Affected Resource(s)/Data Source(s)

azurerm_storage_account

### Terraform Configuration Files

```hcl
resource "azurerm_storage_account" "storage_account" {
  name                     = var.storage_account_name
  location                 = var.resource_group_location
  resource_group_name      = var.resource_group_name
  account_kind             = var.account_kind
  account_tier             = var.account_tier
  account_replication_type = var.replication_type
  min_tls_version          = var.min_tls_version
  nfsv3_enabled            = var.nfsv3_enabled == true ? var.nfsv3_enabled : null
  is_hns_enabled           = var.nfsv3_enabled || var.sftp_enabled ? true : var.is_hns_enabled
  sftp_enabled             = var.sftp_enabled == true ? var.sftp_enabled : null

  blob_properties {
    container_delete_retention_policy {
      
      days = var.storage_blob_data_protection.container_delete_retention_policy_in_days
    }
    delete_retention_policy {
      days = var.storage_blob_data_protection.delete_retention_policy_in_days
    }

    # change_feed_enabled      = var.nfsv3_enabled || var.sftp_enabled ? false : var.storage_blob_data_protection.change_feed_enabled
    versioning_enabled       = var.nfsv3_enabled || var.sftp_enabled || var.is_hns_enabled ? false : var.storage_blob_data_protection.versioning_enabled
    last_access_time_enabled = var.nfsv3_enabled || var.sftp_enabled ? false : var.storage_blob_data_protection.last_access_time_enabled
  }
  tags = var.tags
}
```

### Debug Output/Panic Output

```shell
i have enabled Storage accounts should restrict network access policy on my azure subscription and 
terraform apply for azurerm_storage_account is failing to create storage account.
: Failure sending request: StatusCode=403 -- Original Error: Code="RequestDisallowedByPolicy" Message="Resource 'sredl001' was disallowed by policy. Reasons: 'Storage accounts may not be open to the internet.  Exceptions are granted by architect and VP approval only.'. See error details for policy resource IDs." Target="xxxxxx000011" AdditionalInfo=[{"info":{"evaluationDetails":{"evaluatedExpressions":[{"expression":"type","expressionKind":"Field","expressionValue":"Microsoft.Storage/storageAccounts","operator":"Equals","path":"type","result":"True","targetValue":"Microsoft.Storage/storageAccounts"},{"expression":"Microsoft.Storage/storageAccounts/networkAcls.defaultAction","expressionKind":"Field","expressionValue":"Allow","operator":"NotEquals","path":"properties.networkAcls.defaultAction","result":"True","targetValue":"Deny"}],"reason":"Storage accounts may not be open to the internet.  Exceptions are granted by architect and VP approval only."},
```

**Comment by krupakar1329** [comment]
After enabling **Storage accounts should restrict network access policy** on my azure subscription , terraform is not able to create storage account. 
I am creating network rules through **azurerm_storage_account_network_rules** after storage account creation , but its failing at **azurerm_storage_account** level itself

**Comment by phil-bevan** [comment]
Have you tried setting [public_network_access_enabled](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/storage_account#public_network_access_enabled) to false?

**Comment by krupakar1329** [comment]
@phil-bevan yes, still it fails because , the policy checks the **networkAcls.defaultAction** field
`    "policyRule": {
      "if": {
        "allOf": [
          {
            "field": "type",
            "equals": "Microsoft.Storage/storageAccounts"
          },
          {
            "field": "Microsoft.Storage/storageAccounts/networkAcls.defaultAction",
            "notEquals": "Deny"
          }
        ]
      },
      "then": {
        "effect": "[parameters('effect')]"
      }
    },`

**Comment by eparisca** [comment]
@krupakar1329 In theory, the following may allow you to create the resource and then manage the rules separately under `azurerm_storage_account_network_rules`

```
  network_rules {
    default_action             = "Deny"
    virtual_network_subnet_ids = []
  }
  lifecycle {
    ignore_changes = [
      network_rules,
    ]
  }
```

**Comment by krupakar1329** [comment]
for **azurerm_storage_account_network_rules** 
it throws resource already exists error , if we have network_rules section in **azurerm_storage_account**

**Comment by krupakar1329** [comment]
for **azurerm_storage_account_network_rules** 
it throws resource already exists error , if we have network_rules section in
**azurerm_storage_account**

**Comment by don4of4** [comment]
> @krupakar1329 In theory, the following may allow you to create the resource and then manage the rules separately under `azurerm_storage_account_network_rules`
> 
> ```
> resource "azurerm_storage_account" "storage_account" {
>   # ...
> 
>   # Add a dummy rule that puts the storage account within compliance, and only used upon creation
>   network_rules {
>     default_action             = "Deny"
>     virtual_network_subnet_ids = []
>   }
>   # We will manage rules using azurerm_storage_account_network_rules below
>   lifecycle {
>     ignore_changes = [
>       network_rules,
>     ]
>   }
> }
> 
> resource "azurerm_storage_account_network_rules" "example_1" { 
>   # ...
> }
> ```

The documentation at the top of the page says you can’t have two network rules blocks, unfortunately.

**Comment by eparisca** [comment]
> > @krupakar1329 In theory, the following may allow you to create the resource and then manage the rules separately under `azurerm_storage_account_network_rules`
> > ```
> > resource "azurerm_storage_account" "storage_account" {
> >   # ...
> > 
> >   # Add a dummy rule that puts the storage account within compliance, and only used upon creation
> >   network_rules {
> >     default_action             = "Deny"
> >     virtual_network_subnet_ids = []
> >   }
> >   # We will manage rules using azurerm_storage_account_network_rules below
> >   lifecycle {
> >     ignore_changes = [
> >       network_rules,
> >     ]
> >   }
> > }
> > 
> > resource "azurerm_storage_account_network_rules" "example_1" { 
> >   # ...
> > }
> > ```
> 
> The documentation at the top of the page says you can’t have two network rules blocks, unfortunately.

Correct. I'm not discussing having two `network_rules` blocks, only one.

**Comment by jcockles** [comment]
Hi!
Was there any resolution to this problem at all?
I'm working in an environment where storage accounts in our Azure subscription must be created with the Microsoft.Storage/storageAccounts/networkAcls.defaultAction property set to Deny as defined by a policy. I am unable to set this property through Terraform that I am aware of so my Terraform always fails.

**Comment by AErmie** [comment]
We're also experiencing a similar/related issue in a CAF/Enterprise-Scale deployed environment.

There is an Azure Policy Initiative (called "**Public network access should be disabled for PaaS services**") which has a Policy Definition (called "**Storage accounts should disable public network access**").

Within that policy definition, it checks if `publicNetworkAccess` does not equal `Disabled`. But, the Terraform resource for [azurerm_storage_account](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/storage_account) has a property `public_network_access_enabled` as a **true** or **false** value.

I believe this is an issue with the `azurerm_storage_account` Terraform resource, as it's not "translating" the "false" value into "Disabled" (which the policy is checking for).

```json
    "policyRule": {
      "if": {
        "allOf": [
          {
            "field": "type",
            "equals": "Microsoft.Storage/storageAccounts"
          },
          {
            "field": "Microsoft.Storage/storageAccounts/publicNetworkAccess",
            "notEquals": "Disabled"
          }
        ]
      },
      "then": {
        "effect": "[parameters('effect')]"
      }
    }
```

**Comment by dlane-cbts** [comment]
We had this same issue using the new Azure AVM Modules

By setting both of these options for the `azurerm_storage_account`  it was able to get past the policy

```
  public_network_access_enabled = false
  allow_nested_items_to_be_public = false
```
